File: python/d03.py
import sys

with open(f"../{"test-" if "-t" in sys.argv else ""}data/d03.txt", "r") as file:
    banks = [[int(c) for c in line.strip()] for line in file.readlines()]

result = sum(
    max(
        int(f"{bank[i]}{bank[j]}")
        for i in range(len(bank))
        for j in range(i + 1, len(bank))
    )
    for bank in banks
)
print("p1", result)


def greedy(s: list[int], start_idx: int, n: int) -> int:
    # Find a slice of s such that we can pick one, and still have n-1 left to pick
    # Pick the maximum value in this slice, and denote the index
    # We can now only search FROM THIS INDEX + 1 and forwards
    # Repeat until n is 0
    if n == 0:
        return 0
    elif n == 1:
        seq = s[start_idx:]
    else:
        seq = s[start_idx : -(n - 1)]
    max_idx = start_idx + max(enumerate(seq), key=lambda x: x[1])[0]
    max_val = s[max_idx]
    return max_val * 10 ** (n - 1) + greedy(s, max_idx + 1, n - 1)


print("p2", sum(greedy(bank, 0, 12) for bank in banks))

File: rust/src/bin/d03.rs
use anyhow::{Result, anyhow};
use aoc25::day_input;

type Bank = Vec<u8>;

fn main() -> Result<()> {
    let input = day_input!("03");
    let banks = parse_input(input)?;
    println!("part one: {}", part_one(&banks)?);
    println!("part two: {}", part_two(&banks)?);
    Ok(())
}

fn parse_input(input: &str) -> Result<Vec<Bank>> {
    input
        .lines()
        .map(|line| {
            line.chars()
                .map(|c| Ok(c.to_digit(10).ok_or(anyhow!("invalid character"))? as u8))
                .collect::<Result<Bank>>()
        })
        .collect::<Result<Vec<Bank>>>()
}

fn part_one(banks: &[Bank]) -> Result<usize> {
    banks.iter().map(|bank| greedy(bank, 0, 2)).sum()
}

fn part_two(banks: &[Bank]) -> Result<usize> {
    banks.iter().map(|bank| greedy(bank, 0, 12)).sum()
}

fn greedy(bank: &Bank, start_idx: usize, n: usize) -> Result<usize> {
    if n == 0 {
        return Ok(0);
    }
    let stop_idx = bank.len() - n + 1;
    let seq = &bank[start_idx..stop_idx];
    let (mut max_idx, max_val) = seq
        .iter()
        .copied()
        .enumerate()
        .max_by_key(|(i, v)| (*v, std::cmp::Reverse(*i))) // prefer earlier index on tie
        .ok_or(anyhow!("no maximum value found"))?;
    max_idx += start_idx;
    let term = max_val as usize * 10usize.pow((n - 1) as u32);
    Ok(term + greedy(bank, max_idx + 1, n - 1)?)
}
